pkg/pgmodel/model/custom_types.go

```diff
@@ -7,113 +7,232 @@ package model
+			OID:   arrayOfLabelValueArrayOID,
+			Codec: &pgtype.ArrayCodec{ElementType: labelValueArrayType},
+		},
+		{
+			Name:  "_prom_api._label_array",
+			OID:   arrayOfLabelArrayOID,
+			Codec: &pgtype.ArrayCodec{ElementType: labelArrayType},
+		},
+		{
+			Name:  "ps_trace.tag_type",
+			OID:   traceTagTypeOID,
+			Codec: &pgtype.Int2Codec{},
+		},
+		{
+			Name:  "ps_trace.trace_id",
+			OID:   traceIdOID,
+			Codec: &pgtype.UUIDCodec{},
+		},
+		{
+			Name:  "ps_trace._tag_map",
+			OID:   arrayOfTraceTagMapOID,
+			Codec: &pgtype.ArrayCodec{ElementType: traceTagMapType},
+		},
+		{
+			Name:  "ps_trace.status_code",
+			OID:   traceStatusCodeOID,
+			Codec: &pgtype.EnumCodec{},
+		},
+	}
+
+	registeredTypes[key(conn.Config().Config)] = types
+
+	return types, nil
 }
 
-type PgCustomType uint8
+func getRegisteredTypes(cfg pgconn.Config) ([]*pgtype.Type, bool) {
+	registerTypesMux.RLock()
+	defer registerTypesMux.RUnlock()
+	types, ok := registeredTypes[key(cfg)]
+	return types, ok
+}
 
-const (
-	LabelArray PgCustomType = iota
-	LabelValueArray
-)
+func registerCustomPgTypes(m *pgtype.Map, types []*pgtype.Type) {
+	for _, t := range types {
+		m.RegisterType(t)
+	}
+}
 
-func GetCustomTypeOID(t PgCustomType) uint32 {
-	switch t {
-	case LabelArray:
-		labelArrayOIDMux.Lock()
-		defer labelArrayOIDMux.Unlock()
-		if !isLabelArrayOIDSet {
-			panic("label_array oid is not set. This needs to be set first before calling the type.")
+func SliceToArrayOfLabelArray(src [][]int32) ArrayOfLabelArray {
+	a := make(ArrayOfLabelArray, 0, len(src))
+	for _, i := range src {
+		la := make(LabelArray, 0, len(i))
+		for _, j := range i {
+			la = append(la, pgtype.Int4{Int32: j, Valid: true})
 		}
-		return labelArrayOID
-	case LabelValueArray:
-		labelValueArrayOIDMux.Lock()
-		defer labelValueArrayOIDMux.Unlock()
-		if !isLabelValueArrayOIDSet {
-			panic("label_value_array oid is not set.  This needs to be set first before calling the type.")
-		}
-		return labelValueArrayOID
-	default:
-		panic("invalid type")
+		a = append(a, la)
 	}
+	return a
 }
 
-// GetCustomType returns a custom pgtype.
-func GetCustomType(t PgCustomType) *pgtype.ArrayType {
-	switch t {
-	case LabelArray:
-		if !isLabelArrayOIDSet {
-			panic("label_array oid is not set. This needs to be set first before calling the type.")
-		}
-		return pgtype.NewArrayType("prom_api.label_array", GetCustomTypeOID(t), labelArrayTranscoder)
-	case LabelValueArray:
-		if !isLabelValueArrayOIDSet {
-			panic("label_value_array oid is not set.  This needs to be set first before calling the type.")
-		}
-		return pgtype.NewArrayType("prom_api.label_value_array", GetCustomTypeOID(t), labelValueArrayTranscoder)
-	default:
-		panic("invalid type")
+// Wrapper to allow DecodeBinary to reuse the existing array so that a pool is
+// effective
+type ReusableArray[T any] struct {
+	pgtype.FlatArray[T]
+}
+
+func (a *ReusableArray[T]) SetDimensions(dimensions []pgtype.ArrayDimension) error {
```

Comment:
Since I haven't looked close enough into new PGX implementation is this function supposed to be called by `DecodeBinary` ? Do we need to provide `DecodeBinary` implementation?

Reply:
Also a bit confused with combining `pgtype.FlatArray` and `pgtype.ArrayDimension` ? Isn't FlatArray designed to ignore PostgreSQL dimensions?

Reply:
All the arrays are decoded/encoded by an ArrayCodec. The ArrayCodec works with 2 interfaces which are [ArraySetter and ArrayGetter](https://github.com/jackc/pgx/blob/master/pgtype/array_codec.go#L14-L38) which both [Array](https://github.com/jackc/pgx/blob/master/pgtype/array.go#L400-L433) and [FlatArray](https://github.com/jackc/pgx/blob/master/pgtype/array.go#L445-L481) implement.

ArrayDimensions tells how big the array needs to be, and it uses in the SetDimensions function to allocate the underlaying Array, the difference between Array and FlatArray is that the array struct keeps a reference to the dimensions object, while FlatArray doesn't.